Tests/SwiftFusionTests/Inference/NonlinearFactorGraphTests.swift

```diff
@@ -0,0 +1,99 @@
+import SwiftFusion
+import TensorFlow
+import XCTest
+
+final class NonlinearFactorGraphTests: XCTestCase {
+  /// test ATr
+  func testBasicOps() {
+    var fg = NonlinearFactorGraph()
+    
+    let bf1 = BetweenFactor(0, 1, Pose2(0.0,0.0, 0.0))
+    
+    fg += AnyNonlinearFactor(bf1)
+    
+    var val = Values()
+    val.insert(0, AnyDifferentiable(Pose2(1.0, 1.0, 0.0)))
+    val.insert(1, AnyDifferentiable(Pose2(1.0, 1.0, .pi)))
+    
+    let gfg = fg.linearize(val)
+    
+    var vv = VectorValues()
+    
+    vv.insert(0, Tensor<Double>(shape:[3, 1], scalars: [0.0, 0.0, 0.0]))
+    vv.insert(1, Tensor<Double>(shape:[3, 1], scalars: [0.0, 0.0, 0.0]))
+    
+    let expected = Tensor<Double>(shape:[3, 1], scalars: [.pi, 0.0, 0.0])
+    
+    print("gfg = \(gfg)")
+    print("error = \(gfg.residual(vv).norm)")
+    assertEqual((gfg.residual(vv))[0], expected, accuracy: 1e-6)
+  }
+  
+  /// test CGLS iterative solver
+  func testCGLSPose2SLAM() {
+    // Initial estimate for poses
+    let p1T0 = Pose2(Rot2(0.2), Vector2(0.5, 0.0))
+    let p2T0 = Pose2(Rot2(-0.2), Vector2(2.3, 0.1))
+    let p3T0 = Pose2(Rot2(.pi / 2), Vector2(4.1, 0.1))
+    let p4T0 = Pose2(Rot2(.pi), Vector2(4.0, 2.0))
+    let p5T0 = Pose2(Rot2(-.pi / 2), Vector2(2.1, 2.1))
+
+    let map = [p1T0, p2T0, p3T0, p4T0, p5T0]
+
+    var fg = NonlinearFactorGraph()
+    
+    fg += AnyNonlinearFactor(BetweenFactor(1, 0, Pose2(2.0, 0.0, .pi / 2)))
+    fg += AnyNonlinearFactor(BetweenFactor(2, 1, Pose2(2.0, 0.0, .pi / 2)))
+    fg += AnyNonlinearFactor(BetweenFactor(3, 2, Pose2(2.0, 0.0, .pi / 2)))
+    fg += AnyNonlinearFactor(BetweenFactor(4, 3, Pose2(2.0, 0.0, .pi / 2)))
+    fg += AnyNonlinearFactor(PriorFactor(0, Pose2(0.0, 0.0, 0.0)))
+    
+    var val = Values()
+    
+    for i in 0..<5 {
+      val.insert(i, AnyDifferentiable(map[i]))
+    }
+    
+    for _ in 0..<2 {
+      let gfg = fg.linearize(val)
+      
+      let optimizer = CGLS(precision: 1e-6, max_iteration: 500)
+      
+      var dx = VectorValues()
+      
+      for i in 0..<5 {
+        dx.insert(i, Tensor<Double>(shape: [3, 1], scalars: [0, 0, 0]))
+      }
+      
+      optimizer.optimize(gfg: gfg, initial: &dx)
+      
+      for i in 0..<5 {
```

Comment:
I see two possible improvements to the `Values` API.

1. You could make a subscript that does casting, which would allow you to mutate variables without storing them in an intermediate `var`. Something like `val[i, as: Pose2.self].move(along: ...)`.
2. `Values.TangentVector` could be `VectorValues`, which would allow you to replace this whole loop with a single line `val.move(along: dx)`.

Let me know if you want me to help write code for either of these.

Reply:
agree on both!

Reply:
I think (2) will become possible after #37 is merged? Then I think we can first merge this PR and then #37, and another PR for this one.

Reply:
I think there might be some additional work required to enable (2) after #37 is merged -- it might not be totally easy to use `Manifold` with `Values` because `Values` contains some type erased stuff and type erased stuff sometimes needs some extra work to make it play nicely with everything else.

I'll investigate how to get (2) working.

Your suggestion of fixing this in a separate PR so that we don't block this PR sounds good to me!

Reply:
Done in https://github.com/borglab/SwiftFusion/pull/49!